### PR TITLE
feat(api): Add `requireRestart` parameter

### LINKED ISSUE #283: API: Add `requiresRestart` field to `Secret` and `Variable`
## Description

Whenever a configuration changes, we propagate the changes to the consuming applications. Consider a case where `PORT_NUMBER` changes. We would want our CLI to reload the application on the new port. For this, we would like to introduce a new field called `requireRestart` to both the `Secret` and `Variable` entities.

## Solution

- Update the [`Secret`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/prisma/schema.prisma#L360) and [`Variable`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/prisma/schema.prisma#L396) models to store the new field. It should be a boolean value, with the default set to `false`.
- Update [`create.secret.ts`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/secret/dto/create.secret/create.secret.ts) and [`create.variable.ts`](https://github.com/keyshade-xyz/keyshade/blob/develop/apps/api/src/variable/dto/create.variable/create.variable.ts). The fields should be marked optional.
- Update the service classes to reflect the changes.
- In places where we use Redis to emit the changes in configurations (when rolling back or creating a new revision), add another key called `requiresRestart` to the body.
- Include tests

### PR DESCRIPTION
**Pull request by yogesh1801**
## Description

Added requireRestart param

Fixes #283 

## Developer's checklist

- [ ] My PR follows the style guidelines of this project
- [ ] I have performed a self-check on my work

**If changes are made in the code:**

- [ ] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [ ] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [ ] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [ ] I have made the necessary updates to the documentation, or no documentation changes are required.

___

### **PR Type**
Enhancement, Documentation

___

### **Description**
- Added `requireRestart` boolean property to `CreateSecret` and `CreateVariable` classes.
- Integrated `requireRestart` in secret and variable service operations, including creation, update, and rollback.
- Updated Prisma schema and migration scripts to include `requireRestart` field in `Secret` and `Variable` models.
- Added missing newline at the end of `workspace-storage.ts` file.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>create.secret.ts</strong><dd><code>Add `requireRestart` property to `CreateSecret` class</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/secret/dto/create.secret/create.secret.ts
<li>Added <code>requireRestart</code> boolean property to <code>CreateSecret</code> class.<br> <li> Marked <code>requireRestart</code> as optional.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-0ba28109a277c57e22c8bdc661163b0d1f3b4ae8171a89d3fc72cf3e620529cb">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>secret.service.ts</strong><dd><code>Integrate `requireRestart` in secret service operations</code>&nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/secret/service/secret.service.ts
<li>Included <code>requireRestart</code> in secret creation and update logic.<br> <li> Added <code>requireRestart</code> to the payload for secret rollback.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-dee38177617754972e3d3f727d7f1536566d7a784b7ffdda74aa97d6eec4cbc5">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>create.variable.ts</strong><dd><code>Add `requireRestart` property to `CreateVariable` class</code>&nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/variable/dto/create.variable/create.variable.ts
<li>Added <code>requireRestart</code> boolean property to <code>CreateVariable</code> class.<br> <li> Marked <code>requireRestart</code> as optional.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-f00a0e578eb924e62fd50241ef8fb0583302b9737b42c98cf2f41b965b42c5c3">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>variable.service.ts</strong><dd><code>Integrate `requireRestart` in variable service operations</code></dd></summary>
<hr>

apps/api/src/variable/service/variable.service.ts
<li>Included <code>requireRestart</code> in variable creation and update logic.<br> <li> Added <code>requireRestart</code> to the payload for variable rollback.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-db1895137a9c036373abefb8c924150d4f2022fd6d4f5271cc6c95c443bf01cf">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>migration.sql</strong><dd><code>Add `requireRestart` column to database schema</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/prisma/migrations/20240625190757_new/migration.sql
<li>Added <code>requireRestart</code> column to <code>Secret</code> and <code>Variable</code> tables with default <br>value <code>false</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-23c1bdff5c2702c45704e034efc734de1f56cd8105f4123c81015ce327298532">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>schema.prisma</strong><dd><code>Update Prisma schema with `requireRestart` field</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/prisma/schema.prisma
<li>Added <code>requireRestart</code> boolean field to <code>Secret</code> and <code>Variable</code> models with <br>default value <code>false</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-24191263d57c55d02e0cdf394de15225b628d1da20bd826b1541b59b6b02adb4">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Formatting
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>workspace-storage.ts</strong><dd><code>Add newline at end of workspace storage file</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/platform/src/lib/workspace-storage.ts
- Added missing newline at the end of the file.

</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/286/files#diff-3399b4a605e2a48dd3faa057150c25a91b298f00543cc21fc191839f08d236cb">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

**Comment by rajdip-b** [comment]
Hey bro! Glad that you put up the PR. Are you attempting this for FOSS hack? If so, the PR will need to wait. If not, we can start reviewing it asap!

**Comment by yogesh1801** [comment]
@rajdip-b this PR is my attempt to get familiar with the codebase, will attempt fosshack later

File: apps/api/src/prisma/schema.prisma [hunk 1]
```diff
@@ -365,6 +365,7 @@ model Secret {
   updatedAt DateTime        @updatedAt
   rotateAt  DateTime?
   note      String?
+  requireRestart Boolean        @default(false)
```
Comment (rajdip-b) [inline apps/api/src/prisma/schema.prisma:368]: You would need to format the db once: `pnpm db:format`

**Review by rajdip-b — CHANGES_REQUESTED** [review]

**Review by rajdip-b — CHANGES_REQUESTED** [review]
Update the interface over [here](https://github.com/keyshade-xyz/keyshade/blob/36619b0ab31416f8171198cdb1ccadb525c21fee/apps/api/src/socket/socket.types.ts#L7)

File: apps/api/src/prisma/migrations/20240625190757_new/migration.sql
Comment (rajdip-b) [inline apps/api/src/prisma/migrations/20240625190757_new/migration.sql:1]: Oh missed this one, can you rename the migration folder to something more meaningful? Maybe: `<migration_id>_add_require_restart_field`

**Review by rajdip-b — CHANGES_REQUESTED** [review]

**Review by rajdip-b — APPROVED** [review]

**Pull request merged**